File: server/routes/alias.js
import uuid from 'uuid/v4';
import {pick} from 'lodash';
import Joi from 'joi';
import Boom from 'boom';
import Alias from '../classes/alias';


export default function alias(server) {
  server.route({
    /**
     * GET all alias
     *
     * @header
     * @param {string} JWT token for authentication
     * @return {array} list of alias data
     */
    path: '/api/v3/alias',
    method: 'GET',
    config: {
      auth: {
        strategy: 'token',
      },
    },
    handler: async function(request, reply) {
      const settings = new Alias({server});

      try {
        const data = await settings.getAll(['guid', 'alias', 'ip', 'port', 'mask', 'captureID', 'status']);
        //if (!data || !data.length) {
          //return reply(Boom.notFound('no alias found'));          
        //}
        
        data.sort(function(a, b){
              if(a.guid < b.guid) { return -1; }
              if(a.guid > b.guid) { return 1; }
              return 0;
        });     
                                                                                    
  
        return reply({
          count: data.length,
          data,
        });
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * GET alias by guid
     *
     * @header
     * @param {string} JWT token for authentication
     * @request
     * @param {string} guid
     * @return {array} list of alias data
     */
    path: '/api/v3/alias/{guid}',
    method: 'GET',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        params: {
          guid: Joi.string().min(12).max(46).required(),
        },
      },
    },
    handler: async function(request, reply) {
      const {guid} = request.params;
      const settings = new Alias({server, guid});

      try {
        const data = await settings.get(['guid', 'alias', 'ip', 'port', 'mask', 'captureID', 'status']);
        if (!data || !Object.keys(data).length) {
          return reply(Boom.notFound('no alias found for guid ' + guid));
        }
  
        return reply({
          count: data.length,
          data,
        });
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * Create (POST) a new alias
     *
     * @header
     * @param {string} JWT token for authentication
     * @payload
     * @param {string} alias
     * @param {string} ip
     * @param {number} port
     * @param {number} mask
     * @param {string} captureID
     * @param {boolean} status
     * @return alias guid and success message
     */
    path: '/api/v3/alias',
    method: 'POST',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        payload: {
          alias: Joi.string().min(1).max(40),
          ip: Joi.string().min(6).max(60),
          port: Joi.number(),
          mask: Joi.number(),
          captureID: Joi.string().min(0).max(20),
          status: Joi.boolean(),
        },
      },
    },
    handler: async function(request, reply) {
      const {alias, ip, port, mask, captureID, status} = request.payload;
      const guid = uuid();
      const settings = new Alias({server});

      try {
        await settings.add({guid, alias, ip, port, mask, captureID, status});
        return reply({
          data: guid,
          message: 'successfully created alias',
        }).code(201);
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * Update (PUT) an alias
     *
     * @header
     * @param {string} JWT token for authentication
     * @request
     * @param {string} guid
     * @payload
     * @param {string} alias
     * @param {string} ip
     * @param {number} port
     * @param {number} mask
     * @param {string} captureID
     * @param {boolean} status
     * @return alias guid and success message
     */
    path: '/api/v3/alias/{guid}',
    method: 'PUT',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        params: {
          guid: Joi.string().min(12).max(46).required(),
        },
        payload: {
          alias: Joi.string().min(1).max(40),
          ip: Joi.string().min(6).max(60),
          port: Joi.number(),
          mask: Joi.number(),
          captureID: Joi.string().min(0).max(20),
          status: Joi.boolean(),
        },
      },
      pre: [
        {
          method: async function(request, reply) {
            const {guid} = request.params;
            const settings = new Alias({server, guid});

            try {
              const res = await settings.get(['guid']);
              if (!res || !res.guid || res.guid !== guid) {
                return reply(Boom.notFound(`the alias with id ${guid} was not found`));
              }

              return reply.continue();
            } catch (err) {
              return reply(Boom.serverUnavailable(err));
            }
          },
        },
      ],
    },
    handler: async function(request, reply) {
      const {guid} = request.params;
      const updates = pick(request.payload, ['alias', 'ip', 'port', 'mask', 'captureID', 'status']);

      const settings = new Alias({server, guid});

      try {
        await settings.update(updates);
        return reply({
          data: guid,
          message: 'successfully updated alias',
        }).code(201);
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });

  server.route({
    /**
     * DELETE an alias
     *
     * @header
     * @param {string} JWT token for authentication
     * @request
     * @param {string} guid
     * @return alias guid and success message
     */
    path: '/api/v3/alias/{guid}',
    method: 'DELETE',
    config: {
      auth: {
        strategy: 'token',
      },
      validate: {
        params: {
          guid: Joi.string().min(12).max(46).required(),
        },
      },
      pre: [
        {
          method: async function(request, reply) {
            const {guid} = request.params;
            const settings = new Alias({server, guid});

            try {
              const res = await settings.get(['guid']);
              if (!res || !res.guid || res.guid !== guid) {
                return reply(Boom.notFound(`the alias with id ${guid} were not found`));
              }

              return reply.continue();
            } catch (err) {
              return reply(Boom.serverUnavailable(err));
            }
          },
        },
      ],
    },
    handler: async function(request, reply) {
      const {guid} = request.params;
      const settings = new Alias({server, guid});

      try {
        await settings.delete();
        return reply({
          data: guid,
          message: 'successfully deleted alias',
        }).code(201);
      } catch (err) {
        return reply(Boom.serverUnavailable(err));
      }
    },
  });
};

File: controller/v1/alias.go
package controllerv1

import (
	"fmt"
	"net/http"
	"sort"
	"time"

	"github.com/Jeffail/gabs/v2"
	"github.com/labstack/echo/v4"
	"github.com/sipcapture/homer-app/data/service"
	"github.com/sipcapture/homer-app/model"
	httpresponse "github.com/sipcapture/homer-app/network/response"
	"github.com/sipcapture/homer-app/system/webmessages"
	"github.com/sirupsen/logrus"
)

type AliasController struct {
	Controller
	AliasService *service.AliasService
}

// swagger:route POST /alias alias aliasAddAlias
//
// Adds alias to system
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: AliasStruct
//   in: body
//   description: AliasStruct parameters
//   schema:
//      type: AliasStruct
//   required: true
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
//
// Responses:
//   201: body:AliasCreationSuccessResponse
//   401: body:FailureResponse
func (alc *AliasController) AddAlias(c echo.Context) error {

	aliasObject := model.TableAlias{}
	if err := c.Bind(&aliasObject); err != nil {
		logrus.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, webmessages.UserRequestFormatIncorrect)
	}
	// validate input request body
	if err := c.Validate(aliasObject); err != nil {
		logrus.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, err.Error())
	}

	row, _ := alc.AliasService.Add(&aliasObject)
	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, row)
}

// swagger:route DELETE /alias/{guid} alias aliasDeleteAlias
//
// delete an alias based upon guid
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: guid
//   in: path
//   example: 11111111-1111-1111-1111-111111111111
//   description: guid of the alias to delete
//   required: true
//   type: string
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
//
// Responses:
//   201: body:AliasDeleteSuccessResponse
//   401: body:FailureResponse
func (als *AliasController) DeleteAlias(c echo.Context) error {

	aliasObject := model.TableAlias{}
	aliasObject.GUID = c.Param("guid")
	data, err := als.AliasService.Get(&aliasObject)
	if err != nil {
		reply := gabs.New()
		reply.Set(aliasObject.GUID, "data")
		reply.Set(fmt.Sprintf("the alias with id %s were not found", aliasObject.GUID), "message")
	}

	if err := als.AliasService.Delete(&aliasObject); err != nil {
		reply := gabs.New()
		reply.Set(aliasObject.GUID, "data")
		reply.Set(fmt.Sprintf("the alias with id %s were not found", aliasObject.GUID), "message")
	}

	reply := gabs.New()
	reply.Set(data.GUID, "data")
	reply.Set("successfully deleted alias", "message")

	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, reply.String())
}

// swagger:route PUT /alias/{guid} alias aliasUpdateAlias
//
// Update an existing alias based upon the lookup
// ---
// consumes:
// - application/json
// produces:
// - application/json
// parameters:
// + name: guid
//   in: path
//   example: 11111111-1111-1111-1111-111111111111
//   description: uuid of the Alias to update
//   required: true
//   type: string
// + name: AliasStruct
//   in: body
//   description: AliasStruct parameters
//   schema:
//      type: AliasStruct
//   required: true
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
//
// Responses:
//   201: body:AliasUpdateSuccessResponse
//   401: body:FailureResponse
func (als *AliasController) UpdateAlias(c echo.Context) error {
	aliasObject := model.TableAlias{}
	if err := c.Bind(&aliasObject); err != nil {
		logrus.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, webmessages.UserRequestFormatIncorrect)
	}
	// validate input request body
	if err := c.Validate(aliasObject); err != nil {
		logrus.Error(err.Error())
		return httpresponse.CreateBadResponse(&c, http.StatusBadRequest, err.Error())
	}
	aliasObject.GUID = c.Param("guid")
	data, err := als.AliasService.Get(&aliasObject)
	if err != nil {
		reply := gabs.New()
		reply.Set(aliasObject.GUID, "data")
		reply.Set(fmt.Sprintf("the alias with id %s were not found", aliasObject.GUID), "message")
	}

	aliasObject.CreateDate = time.Now()
	aliasObject.Id = data.Id
	if err := als.AliasService.Update(&aliasObject); err != nil {
		reply := gabs.New()
		reply.Set(aliasObject.GUID, "data")
		reply.Set(fmt.Sprintf("the alias with id %s were not found", aliasObject.GUID), "message")
	}

	reply := gabs.New()
	reply.Set(data.GUID, "data")
	reply.Set("successfully updated alias", "message")

	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, reply.String())
}

// swagger:route GET /alias alias aliasGetAllAlias
//
// Get all aliases
// ---
// produces:
// - application/json
// Security:
// - bearer: []
//
// SecurityDefinitions:
// bearer:
//      type: apiKey
//      name: Authorization
//      in: header
// responses:
//   200: body:AliasStructList
func (alc *AliasController) GetAllAlias(c echo.Context) error {

	alias, _ := alc.AliasService.GetAll()

	sort.Slice(alias[:], func(i, j int) bool {
		return alias[i].GUID < alias[j].GUID
	})

	reply := gabs.New()
	reply.Set(alias, "data")

	//reply, _ := json.Marshal(alias)

	return httpresponse.CreateSuccessResponse(&c, http.StatusCreated, reply.String())
}
